core/domain/stats_jobs_one_off.py

```diff
@@ -507,6 +868,219 @@ def reduce(exp_id, stringified_values):
+                exp_id, exp_version, 'num_actual_starts_v2',
+                num_actual_starts_v2)
+
+            # Number of starts must never be less than the number of
+            # completions.
+            if num_starts_v2 < num_completions_v2:
+                yield ('Completions > starts: exp_id:%s version:%s %s > %s' % (
+                    exp_id, exp_version, num_completions_v2, num_starts_v2),)
+
+            # Number of actual starts must never be less than the number of
+            # completions.
+            if num_actual_starts_v2 < num_completions_v2:
+                yield (
+                    'Completions > actual starts: exp_id:%s '
+                    'version:%s %s > %s' % (
+                        exp_id, exp_version, num_completions_v2,
+                        num_actual_starts_v2),)
+
+            # Number of starts must never be less than the number of actual
+            # starts.
+            if num_starts_v2 < num_actual_starts_v2:
+                yield (
+                    'Actual starts > starts: exp_id:%s '
+                    'version:%s %s > %s' % (
+                        exp_id, exp_version, num_actual_starts_v2,
+                        num_starts_v2),)
+
+            for state_name in exp_stats['state_stats_mapping']:
+                state_stats = exp_stats['state_stats_mapping'][state_name]
+
+                total_answers_count_v2 = state_stats['total_answers_count_v2']
+                useful_feedback_count_v2 = state_stats[
+                    'useful_feedback_count_v2']
+                total_hit_count_v2 = state_stats['total_hit_count_v2']
+                first_hit_count_v2 = state_stats['first_hit_count_v2']
+                num_completions_v2 = state_stats['num_completions_v2']
+
+                StatisticsAuditV2.require_non_negative(
+                    exp_id, exp_version, 'total_answers_count_v2',
+                    total_answers_count_v2, state_name=state_name)
+                StatisticsAuditV2.require_non_negative(
+                    exp_id, exp_version, 'useful_feedback_count_v2',
+                    useful_feedback_count_v2, state_name=state_name)
+                StatisticsAuditV2.require_non_negative(
+                    exp_id, exp_version, 'total_hit_count_v2',
+                    total_hit_count_v2, state_name=state_name)
+                StatisticsAuditV2.require_non_negative(
+                    exp_id, exp_version, 'first_hit_count_v2',
+                    first_hit_count_v2, state_name=state_name)
+                StatisticsAuditV2.require_non_negative(
+                    exp_id, exp_version, 'num_completions_v2',
+                    num_completions_v2, state_name=state_name)
+
+                # The total number of answers submitted can never be less than
+                # the number of submitted answers for which useful feedback was
+                # given.
+                if total_answers_count_v2 < useful_feedback_count_v2:
```

Comment:
maybe also check solution count is less than hit count? (I think solution counts are new in v2.)

In general you don't seem to have checks for solution counts (nonnegative etc.)?

Reply:
Done.